src/aquarium-optimized/dawn/ContextDawn.cpp

```diff
@@ -743,14 +786,47 @@ void ContextDawn::reallocResource(int preTotalInstance,
     }
 }
 
+void ContextDawn::MapWriteCallback(DawnBufferMapAsyncStatus status,
+                                   void *data,
+                                   uint64_t,
+                                   void *userdata)
+{
+    ASSERT(status == DAWN_BUFFER_MAP_ASYNC_STATUS_SUCCESS);
+    ASSERT(data != nullptr);
+    ContextDawn *contextDawn = static_cast<ContextDawn *>(userdata);
+
+    int size                = contextDawn->mCurTotalInstance;
+    contextDawn->mappedData = data;
+    memcpy(contextDawn->mappedData, contextDawn->fishPers,
+           sizeof(FishPer) * contextDawn->mCurTotalInstance);
+
+    // Create a command buffer that copy from the staging buffer to dest buffer.
+    dawn::CommandBuffer command = contextDawn->copyBufferToBuffer(
+        contextDawn->stagingBuffer, 0, contextDawn->fishPersBuffer, 0, sizeof(FishPer) * size);
+    contextDawn->mCommandBuffers.emplace_back(command);
+}
+
+void ContextDawn::WaitABit()
+{
+    mDevice.Tick();
+
+    utils::USleep(100);
+}
+
 void ContextDawn::updateAllFishData()
 {
-    setBufferData(fishPersBuffer, 0, sizeof(FishPer) * mCurTotalInstance, fishPers);
+    // SetSubData is a sync method to upload data and the API will be eliminated in Dawn soon.

```

Comment:
Still supported. See https://github.com/gpuweb/gpuweb/pull/418

Reply:
@bbernhar When updating fish uniforms by map buffer aync, ContextDawn::WaitABit is called to wait for the commands of callback function be executed. The application also execute other commands in the function flush. So 2 commit will be called in one frame. This impact on performance. Should I still use a sync method to update the fish uniforms in the render loop? For example, createBufferMapped during initializaton. I can reserve both sync and async method as well.

Reply:
> Should I still use a sync method to update the fish uniforms in the render loop

Yup. Since it has to block or wait to upload the data, anyway.  If the uniforms are dynamic (modified each frame) then SetSubData is the way-to-go. Buffer mapping is ideal for static data (global lights, camera, etc) not per instance.